Intersect.Client.Core/Interface/Game/Spells/SpellsWindow.cs

```diff
@@ -1,191 +1,94 @@
-            Items[i].Pnl.IsHidden = spell == null || Items[i].IsDragging;
-            if (spell != null)
-            {
-                Items[i].Update();
-            }
+            Items[slotIndex].Update();
         }
     }
 
     private void InitItemContainer()
     {
-        for (var i = 0; i < Options.Instance.Player.MaxSpells; i++)
+        for (var slotIndex = 0; slotIndex < Options.Instance.Player.MaxSpells; slotIndex++)
         {
-            Items.Add(new SpellItem(this, i));
-            Items[i].Container = new ImagePanel(mItemContainer, "Spell");
-            Items[i].Setup();
-
-            Items[i].Container.LoadJsonUi(GameContentManager.UI.InGame, Graphics.Renderer.GetResolutionString());
-
-            var xPadding = Items[i].Container.Margin.Left + Items[i].Container.Margin.Right;
-            var yPadding = Items[i].Container.Margin.Top + Items[i].Container.Margin.Bottom;
-            Items[i]
-                .Container.SetPosition(
-                    i %
-                    (mItemContainer.Width / (Items[i].Container.Width + xPadding)) *
-                    (Items[i].Container.Width + xPadding) +
-                    xPadding,
-                    i /
-                    (mItemContainer.Width / (Items[i].Container.Width + xPadding)) *
-                    (Items[i].Container.Height + yPadding) +
-                    yPadding
-                );
+            Items.Add(new SpellItem(this, _slotContainer, slotIndex, _contextMenu));
         }
-    }
 
-    public void Show()
-    {
-        mSpellWindow.IsHidden = false;
+        PopulateSlotContainer.Populate(_slotContainer, Items);
     }
 
-    public bool IsVisible()
+    public override void Hide()
     {
-        return !mSpellWindow.IsHidden;
-    }
-
-    public void Hide()
-    {
-        mContextMenu?.Close();
-        mSpellWindow.IsHidden = true;
+        _contextMenu?.Close();
+        base.Hide();
     }
 
     public FloatRect RenderBounds()
     {
         var rect = new FloatRect()
         {
-            X = mSpellWindow.ToCanvas(new Point(0, 0)).X -
-                (Items[0].Container.Padding.Left + Items[0].Container.Padding.Right) / 2,
-            Y = mSpellWindow.ToCanvas(new Point(0, 0)).Y -
-                (Items[0].Container.Padding.Top + Items[0].Container.Padding.Bottom) / 2,
-            Width = mSpellWindow.Width + Items[0].Container.Padding.Left + Items[0].Container.Padding.Right,
-            Height = mSpellWindow.Height + Items[0].Container.Padding.Top + Items[0].Container.Padding.Bottom
+            X = ToCanvas(new Point(0, 0)).X -
+                (Items[0].Padding.Left + Items[0].Padding.Right) / 2,
+            Y = ToCanvas(new Point(0, 0)).Y -
+                (Items[0].Padding.Top + Items[0].Padding.Bottom) / 2,
+            Width = Width + Items[0].Padding.Left + Items[0].Padding.Right,
+            Height = Height + Items[0].Padding.Top + Items[0].Padding.Bottom
```

Comment:
`ToCanvas()` and `Items[0]` can be cached, and `Items[0]` needs to be safety checked so we don't have a crash.

The `/ 2`s should probably be `/ 2f` to avoid loss of fraction